src/toolkits/activity_classification/ac_data_iterator.cpp

```diff
@@ -132,152 +132,291 @@ variant_map_type _activity_classifier_prepare_data_impl(const gl_sframe &data,
+                       << std::endl;
+  }
 
-            if (curr_window_targets.size() == static_cast<size_t>(prediction_window)) {
-                auto target_val = vec_mode(curr_window_targets);
-                curr_chunk_targets.push_back(target_val);
-                curr_window_targets.clear();
-            }
-        }
-        // Check if the aggregated chunk data has reached the maximal chunk length, and finalize
-        // the chunk processing.
-        if (curr_chunk_features.size() == static_cast<size_t>(feature_size)) {
-            finalize_chunk(curr_chunk_features,
-                           curr_chunk_targets,
-                           curr_window_targets,
-                           curr_session_id,
-                           output_writer,
-                           chunk_size,
-                           feature_size,
-                           predictions_in_chunk,
-                           use_target);
-        }
+  // Build a dict of the column order by column name, to later access within
+  // the iterator
+  auto column_index_map = generate_column_index_map(data.column_names());
+  size_t chunk_length = prediction_window * predictions_in_chunk;
+
+  flex_vec curr_chunk_targets;
+  flex_vec curr_chunk_features;
+
+  flexible_type last_session_id = data[session_id][0];
+  size_t number_of_sessions = 0;
+
+  // Prepare an output SFrame writer, that will write a new SFrame in the
+  // converted batch-processing ready format.
+  std::vector<std::string> output_column_names = {"features", "chunk_len",
+                                                  "session_id"};
+  std::vector<flex_type_enum> output_column_types = {flex_type_enum::VECTOR,
+                                                     flex_type_enum::INTEGER,
+                                                     data[session_id].dtype()};
+  if (use_target) {
+    output_column_names.insert(output_column_names.end(), {"target"});
+    output_column_types.insert(output_column_types.end(),
+                               {flex_type_enum::VECTOR});
+  }
 
-        time_t now = time(0);
-        if (verbose && difftime(now, last_print_time) > 10) {
-            logprogress_stream << "Pre-processing: " << std::setw(3) << (100 * processed_lines / data.size()) << "% complete"  << std::endl;
-            last_print_time = now;
+  gl_sframe_writer output_writer(output_column_names, output_column_types, 1);
+
+  if (verbose) {
+    logprogress_stream << "Using sequences of size " << chunk_length
+                       << " for model creation." << std::endl;
+  }
+
+  time_t last_print_time = time(0);
+  size_t processed_lines = 0;
+
+  // Iterate over the user data. The features and targets are aggregated, and
+  // handled whenever a the ending of a session is reached.
+  size_t chunk_size = 0;
+  for (const auto &line : data.range_iterator()) {
+    auto curr_session_id = line[column_index_map[session_id]];
```

Comment:
May as well use `flexible_type` instead of `auto`
It’s not much longer and makes it clearer to the reader what the type is